charmcraft/commands/init.py

```diff
@@ -0,0 +1,112 @@
+
+    def fill_parser(self, parser):
+        parser.add_argument(
+            "-d", "--directory", type=Path, default=Path("."),
+            help="the directory to initialize. Must not exist or be empty; defaults to '.'")
+        parser.add_argument(
+            "--name", type=str,
+            help="the name of the project; defaults to the directory name")
+
+        parser.add_argument(
+            "--author", type=str,
+            help="the author of the project; defaults to the directory name;"
+            " defaults to the current user's name as present in the GECOS field.")
+
+    def run(self, args):
+        args.directory = args.directory.resolve()
+        if args.directory.exists():
+            if not args.directory.is_dir():
+                raise CommandError("{} is not a directory".format(args.directory))
+            for _ in args.directory.iterdir():
+                raise CommandError("{} is not empty".format(args.directory))
+            logger.debug("Using existing project directory '%s'", args.directory)
+        else:
+            logger.debug("Creating project directory '%s'", args.directory)
+            args.directory.mkdir()
+
+        if args.author is None:
+            gecos = pwd.getpwuid(os.getuid()).pw_gecos.split(',', 1)[0]
+            if not gecos:
+                raise CommandError("Author not given, and nothing in GECOS field")
+            logger.debug("Setting author to %r from GECOS field", gecos)
+            args.author = gecos
+
+        if not args.name:
+            args.name = args.directory.name
+            logger.debug("Set project name to '%s'", args.name)
+
+        context = {
+            "name": args.name,
+            "author": args.author,
+            "year": date.today().year,
+            "class_name": "".join(re.split(r"\W+", args.name.title())) + "Charm",
+        }
+
+        env = Environment(
+            loader=PackageLoader('charmcraft', 'templates/init'),
+            autoescape=False,            # no need to escape things here :-)
+            keep_trailing_newline=True,  # they're not text files if they don't end in newline!
+            optimized=False,             # optimization doesn't make sense for one-offs
+            undefined=StrictUndefined)   # fail on undefined
+
+        # bare minimum:
+        #  🗷 a req.txt w/ops,
+        #  ☐ src/charm.py w/structure & xxx & a minimum hook & stored state
+        #  ☐ a basic fixme test which pulls in the harness and tests that the minimum thing works
+        #  🗷 a run_tests script & flake8
+        #  🗷 a req-dev.txt w/flake8 (etc?)
+        # maybe a templates directory, later :-)
+
+        _xxx_rx = re.compile("XXX: (.*)")
+        executables = ["run_tests", "src/charm.py"]
+        for template_name in env.list_templates():
+            if template_name.endswith(".pyc") or template_name.endswith("~"):
+                # would be nice if we could stop from shipping these 🤷
```

Comment:
Why would we have these? Where is it reading the templates from that we are actually executing the template code? Isn't it illegal code since it hasn't expanded the template variables?

Reply:
(discussed in the standup, but: the last step of install is a compileall pass; pytest also aggressively compiles things during discovery but i dissuaded it)